ohci1394, sbp2: fix "scsi_add_device failed" with PL-3507 based devices

Re-enable posted writes for sbp2's status FIFO.
Besides bringing back a very minor bandwidth tweak from Linux 2.6.15.x
and older, this also fixes an interoperability regression since 2.6.16:
http://bugzilla.kernel.org/show_bug.cgi?id=6356
(sbp2: scsi_add_device failed. IEEE1394 HD is not working anymore.)

Signed-off-by: Stefan Richter <stefanr@s5r6.in-berlin.de>
---
Suggested for -stable because many 1394/USB combo disk enclosures are
built around the Prolific PL-3507 bridge.

The ugly hardwired PhyUpperBound which this patch (re-)introduces into
sbp2 will be eliminated by a subsequent patchset (which has already been
sent to linux1394-devel).

Index: linux-2.6.16.5/drivers/ieee1394/ohci1394.c
===================================================================
--- linux-2.6.16.5.orig/drivers/ieee1394/ohci1394.c	2006-04-19 20:43:56.000000000 +0200
+++ linux-2.6.16.5/drivers/ieee1394/ohci1394.c	2006-04-19 20:49:49.000000000 +0200
@@ -2525,7 +2525,7 @@ static irqreturn_t ohci_irq_handler(int 
 			if (phys_dma) {
 				reg_write(ohci,OHCI1394_PhyReqFilterHiSet, 0xffffffff);
 				reg_write(ohci,OHCI1394_PhyReqFilterLoSet, 0xffffffff);
-				reg_write(ohci,OHCI1394_PhyUpperBound, 0xffff0000);
+				reg_write(ohci,OHCI1394_PhyUpperBound, 0x01000000);
 			} else {
 				reg_write(ohci,OHCI1394_PhyReqFilterHiSet, 0x00000000);
 				reg_write(ohci,OHCI1394_PhyReqFilterLoSet, 0x00000000);
Index: linux-2.6.16.5/drivers/ieee1394/sbp2.c
===================================================================
--- linux-2.6.16.5.orig/drivers/ieee1394/sbp2.c	2006-04-19 20:43:56.000000000 +0200
+++ linux-2.6.16.5/drivers/ieee1394/sbp2.c	2006-04-19 20:53:07.000000000 +0200
@@ -754,11 +754,16 @@ static struct scsi_id_instance_data *sbp
 
 	/* Register the status FIFO address range. We could use the same FIFO
 	 * for targets at different nodes. However we need different FIFOs per
-	 * target in order to support multi-unit devices. */
+	 * target in order to support multi-unit devices.
+	 * The FIFO is located out of the local host controller's physical range
+	 * but, if possible, within the posted write area. Status writes will
+	 * then be performed as unified transactions. This slightly reduces
+	 * bandwidth usage, and some Prolific based devices seem to require it.
+	 */
 	scsi_id->status_fifo_addr = hpsb_allocate_and_register_addrspace(
 			&sbp2_highlevel, ud->ne->host, &sbp2_ops,
 			sizeof(struct sbp2_status_block), sizeof(quadlet_t),
-			~0ULL, ~0ULL);
+			0x010000000000ULL, CSR1212_ALL_SPACE_END);
 	if (!scsi_id->status_fifo_addr) {
 		SBP2_ERR("failed to allocate status FIFO address range");
 		goto failed_alloc;
